components/scream/src/physics/p3/p3_inputs_initializer.cpp

```diff
@@ -45,88 +47,251 @@ add_field (const field_type &f, const field_type& f_ref,
+  auto h_ni_activated     = Kokkos::create_mirror_view(d_ni_activated   ); 
+  auto h_nc_nuceat_tend   = Kokkos::create_mirror_view(d_nc_nuceat_tend ); 
+  auto h_pmid             = Kokkos::create_mirror_view(d_pmid           ); 
+  auto h_dp               = Kokkos::create_mirror_view(d_dp             ); 
+  auto h_zi               = Kokkos::create_mirror_view(d_zi             ); 
+  auto h_qv_prev          = Kokkos::create_mirror_view(d_qv_prev        ); 
+  auto h_T_prev           = Kokkos::create_mirror_view(d_T_prev         ); 
+  auto h_qv               = Kokkos::create_mirror_view(d_qv             ); 
+  auto h_qc               = Kokkos::create_mirror_view(d_qc             ); 
+  auto h_qr               = Kokkos::create_mirror_view(d_qr             ); 
+  auto h_qi               = Kokkos::create_mirror_view(d_qi             ); 
+  auto h_qm               = Kokkos::create_mirror_view(d_qm             ); 
+  auto h_nc               = Kokkos::create_mirror_view(d_nc             ); 
+  auto h_nr               = Kokkos::create_mirror_view(d_nr             ); 
+  auto h_ni               = Kokkos::create_mirror_view(d_ni             ); 
+  auto h_bm               = Kokkos::create_mirror_view(d_bm             ); 
+  auto h_nccn_prescribed  = Kokkos::create_mirror_view(d_nccn_prescribed); 
+  auto h_inv_qc_relvar    = Kokkos::create_mirror_view(d_inv_qc_relvar  ); 
+  // TODO: Delete eventually, should instead be set in run interface: 
+  auto h_th_atm           = Kokkos::create_mirror_view(d_th_atm         ); 
+  auto h_dz               = Kokkos::create_mirror_view(d_dz             ); 
+  auto h_exner            = Kokkos::create_mirror_view(d_exner          ); 
+  auto h_cld_frac_l       = Kokkos::create_mirror_view(d_cld_frac_l     ); 
+  auto h_cld_frac_i       = Kokkos::create_mirror_view(d_cld_frac_i     ); 
+  auto h_cld_frac_r       = Kokkos::create_mirror_view(d_cld_frac_r     ); 
 
-  // Deep copy back to device
-  Kokkos::deep_copy(d_q,h_q);
-  Kokkos::deep_copy(d_T,h_T);
-  Kokkos::deep_copy(d_ast,h_ast);
-  Kokkos::deep_copy(d_ni_activated,h_ni_activated);
-  Kokkos::deep_copy(d_nc_nuceat_tend,h_nc_nuceat_tend);
-  Kokkos::deep_copy(d_pmid,h_pmid);
-  Kokkos::deep_copy(d_dpres,h_dpres);
-  Kokkos::deep_copy(d_zi,h_zi);
-  Kokkos::deep_copy(d_qv_prev,h_qv_prev);
-  Kokkos::deep_copy(d_t_prev,h_t_prev);
-
-  // If we are in charge of init-ing FQ as well, init it to 0.
-  if (m_fields.count("FQ")==1) {
-    // Init FQ to 0
-    auto d_FQ = m_fields.at("FQ").get_view();
-    Kokkos::deep_copy(d_FQ,Real(0));
+ // Initalize from text file 
+  std::ifstream fid("p3_init_vals.txt", std::ifstream::in);
+  std::string tmp_line;
+  int icol_in_max = 0;
+  while(getline(fid,tmp_line))
+  {
+    std::stringstream s(tmp_line);
+    std::string field;
+    std::vector<Real> field_vals;
+    while (getline(s,field,' '))
+    {
+      field_vals.push_back(std::stod(field));
+    }
+    int icol  = (int)field_vals[0];
```

Comment:
You could avoid having `field_vals`, and just do 

```
int  icol, k, ipack, ivec;
s >> icol >> k;
ipack = k / n;
ivec = k % n;

s >> h_qv(icol,ipack)[ivec];
s >> h_th_atm(icol,ipack)[ivec];
s >> h_pmid(icol,ipack)[ivec];
...
```
`std::stringstream` is designed precisely to support this kind of (generic) input/output operations.

Note: if you need to skip some values, you can always do 
```
std::string skip;
s >> skip >> skip >> h_qv(icol,ipack)[ivec];
...
```

Reply:
excellent, thank you for the suggestion.  I've cleaned this up and done this.